encoding/thrift/inbound.go

```diff
@@ -101,58 +75,100 @@ func (t thriftUnaryHandler) Handle(ctx context.Context, treq *transport.Request,
 	}
 
 	if err := call.WriteToResponse(rw); err != nil {
+		// not reachable
 		return err
 	}
 
-	err = proto.EncodeEnveloped(wire.Envelope{
-		Name:  res.Body.MethodName(),
-		Type:  res.Body.EnvelopeType(),
-		SeqID: envelope.SeqID,
-		Value: value,
-	}, rw)
-	if err != nil {
+	if err = responder.EncodeResponse(value, wire.Reply, rw); err != nil {
 		return errors.ResponseBodyEncodeError(treq, err)
 	}
 
 	return nil
 }
 
-// TODO(apb): reduce commonality between Handle and HandleOneway
-
 func (t thriftOnewayHandler) HandleOneway(ctx context.Context, treq *transport.Request) error {
-	if err := errors.ExpectEncodings(treq, Encoding); err != nil {
-		return err
-	}
-
-	ctx, call := encodingapi.NewInboundCall(ctx)
-	if err := call.ReadFromRequest(treq); err != nil {
-		return err
-	}
-
 	buf := bufferpool.Get()
 	defer bufferpool.Put(buf)
-	if _, err := buf.ReadFrom(treq.Body); err != nil {
-		return err
-	}
 
-	// We disable enveloping if either the client or the transport requires it.
-	proto := t.Protocol
-	if !t.Enveloping {
-		proto = disableEnvelopingProtocol{
-			Protocol: proto,
-			Type:     wire.OneWay, // we only decode oneway requests
-		}
-	}
-
-	envelope, err := proto.DecodeEnveloped(bytes.NewReader(buf.Bytes()))
+	ctx, _, reqValue, _, err := decodeRequest(ctx, buf, treq, wire.OneWay, t.Protocol, t.Enveloping)
 	if err != nil {
-		return errors.RequestBodyDecodeError(treq, err)
+		return err
 	}
 
-	if envelope.Type != wire.OneWay {
-		return errors.RequestBodyDecodeError(
-			treq, errUnexpectedEnvelopeType(envelope.Type))
+	return t.OnewayHandler(ctx, reqValue)
+}
+
+// decodeRequest is a utility shared by Unary and Oneway handlers, to decode
+// the request, regardless of enveloping.
+func decodeRequest(
+	parentCtx context.Context,
+	// buf is a byte buffer from the buffer pool, that will be released back to
+	// the buffer pool by the caller after it is finished with the decoded
+	// request payload. Thrift requests read sets, maps, and lists lazilly.
+	buf *bufferpool.Buffer,
+	treq *transport.Request,
+	// reqEnvelopeType indicates the expected envelope type, if an envelope is
+	// present.
+	reqEnvelopeType wire.EnvelopeType,
+	// proto is the encoding protocol (e.g., Binary) or an
+	// EnvelopeAgnosticProtocol (e.g., EnvelopeAgnosticBinary)
+	proto protocol.Protocol,
+	// enveloping indicates that requests must be enveloped, used only if the
+	// protocol is not envelope agnostic.
+	enveloping bool,
+) (
+	// ctx is a context including the inbound call.
```

Comment:
This is a lot of returns for a function, and we end up using a lot of naked returns which makes things a little harder to reason about. E.g., if we had a nested case where we accidentally did `err := `, then we'd set the wrong `err` variable, and that would break.

Is there parts that can be broken up here? 

E.g., right now we take a `ctx` in, and return a `ctx`. But looking at the use of the context, it seems like that could be a completely separate method, and we could just pass in the `call` to this method instead. That would remove two returns (since `call` is passed in, and `ctx` is not modified, so no need to return).

Once we do that, I'd prefer to avoid naked returns, and return directly instead. That would make a few things cleaner.

E.g., we could replace the var declaration with just `envelope, err := `
```go
		var envelope wire.Envelope
		envelope, err = proto.DecodeEnveloped(reader)
		if err != nil { [...] }
```

```
		reqValue, responder, err = agnosticProto.DecodeRequest(reqEnvelopeType, reader)
```

That would be a return, and since we return, we could even reduce nesting for the "long" (e.g., enveloped) case.

Reply:
There are things I can do to reduce the scope of this function, but each involves duplicating code between the two call-sites. Bare returns help this function avoid a bunch of variable name shadowing—I can make an A/B without bare returns for comparison.

Reply:
Here is a variation without bare returns https://github.com/yarpc/yarpc-go/compare/re-envelope-agnostic-thrift-b#diff-6a847d4732c88b341210714b4c091491R101

Reply:
Added a commit that further decomposes decodeRequest without popping it entirely. I’d like to avoid expanding the two decodeRequest call sites.